platform/linux-generic/test/validation/api/ml/ml_linux.c

```diff
@@ -0,0 +1,1259 @@
+	CU_ASSERT_NOT_EQUAL(compl, ODP_ML_COMPL_INVALID);
+
+	if (load_result)
+		ret = odp_ml_compl_load_result(compl, load_result);
+	else
+		ret = odp_ml_compl_run_result(compl, run_result);
+
+	CU_ASSERT_EQUAL(ret, 0);
+
+	return ret;
+}
+
+/* Test asynchronous model loading in ODP_ML_COMPL_MODE_EVENT mode */
+static void test_ml_model_load_async_event(void)
+{
+	int ret;
+	odp_ml_compl_t compl;
+	odp_ml_load_result_t result;
+	odp_ml_compl_param_t compl_param;
+	int dummy = 6;
+	void *user_ptr = &dummy;
+
+	compl = odp_ml_compl_alloc(global.compl_pool);
+	CU_ASSERT_PTR_NOT_EQUAL_FATAL(compl, ODP_ML_COMPL_INVALID);
+
+	odp_ml_compl_param_init(&compl_param);
+	compl_param.mode = ODP_ML_COMPL_MODE_EVENT;
+	compl_param.event = odp_ml_compl_to_event(compl);
+	compl_param.queue = global.queue;
+	compl_param.user_ptr = user_ptr;
+
+	ret = odp_ml_model_load_start(global.ml_model, &compl_param);
+	CU_ASSERT_EQUAL(ret, 0);
+
+	/* Return when odp_ml_model_load_start() failed */
+	if (ret) {
+		odp_ml_compl_free(compl);
+		ODPH_ERR("ML model odp_ml_model_load_start() failed\n");
+		return;
+	}
+
+	/* Run event scheduler to find the ml completion event and verify it */
+	if (get_result_from_ml_compl_event(&result, NULL)) {
+		odp_ml_compl_free(compl);
+		ret = odp_ml_model_unload(global.ml_model, NULL);
+		CU_ASSERT_EQUAL(ret, 0);
+		return;
+	}
+
+	CU_ASSERT_EQUAL(result.error_code, 0);
+	CU_ASSERT_PTR_EQUAL(result.user_ptr, user_ptr);
+
+	/* Model load does not modify data in user_ptr */
+	if (result.user_ptr)
+		CU_ASSERT_EQUAL(*(int *)result.user_ptr, dummy);
+
+	/* Free the event */
+	odp_ml_compl_free(compl);
+
+	CU_ASSERT_EQUAL_FATAL(odp_ml_model_unload(global.ml_model, NULL), 0);
+}
+
+static void test_ml_model_unload(void)
+{
+	int ret;
+	odp_ml_load_result_t result;
+
+	/* Load model in order to test unload */
+	ret = odp_ml_model_load(global.ml_model, NULL);
+	CU_ASSERT_EQUAL_FATAL(ret, 0);
+
+	/* Test when result is not NULL */
+	ret = odp_ml_model_unload(global.ml_model, &result);
+	CU_ASSERT_EQUAL(ret, 0);
+	CU_ASSERT_EQUAL(result.error_code, 0);
+}
+
+/* Test asynchronous model unloading in ODP_ML_COMPL_MODE_POLL mode */
+static void test_ml_model_unload_async_poll(void)
+{
+	int ret;
+	odp_ml_compl_param_t compl_param;
+	int dummy = 6;
+	void *user_ptr = &dummy;
+	uint32_t compl_id = 1;
+	odp_ml_load_result_t result;
+	uint64_t wait_ns = 500 * ODP_TIME_MSEC_IN_NS;
+
+	/* Load model in order to test unload */
+	ret = odp_ml_model_load(global.ml_model, NULL);
+	CU_ASSERT_EQUAL_FATAL(ret, 0);
+
+	memset(&result, 0, sizeof(result));
+	odp_ml_compl_param_init(&compl_param);
+	compl_param.mode = ODP_ML_COMPL_MODE_POLL;
+	compl_param.compl_id = compl_id;
+	compl_param.user_ptr = user_ptr;
+
+	ret = odp_ml_model_unload_start(global.ml_model, &compl_param);
+	CU_ASSERT_EQUAL(ret, 0);
+
+	/* Return when odp_ml_model_unload_start() failed */
+	if (ret) {
+		ret = odp_ml_model_unload(global.ml_model, NULL);
+		CU_ASSERT_EQUAL_FATAL(ret, 0);
+		return;
```

Comment:
If unload start fails, it can be fatal assert. No need to try recovery with sync API.